### PR TITLE
Setup project with OIDC authentication 

### PR DESCRIPTION
**Pull request by mingfeng**
This PR added OIDC authenticated related components, hooks, and other tools. The `auth` module is copied from https://github.com/City-of-Helsinki/example-ui-profile with some modifications. The private pages are protected with `makePrivate` HOC function.

Refs: PV-213

### CODE REVIEW

File: src/auth/hooks.ts [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+  const client = useClient();
+  const tokens = client.isAuthenticated() ? client.getApiTokens() : undefined;
+  const hasTokens = tokens && Object.keys(tokens).length;
+  const [apiTokens, setApiTokens] = useState<JWTPayload | undefined>(
+    hasTokens ? tokens : undefined
+  );
+
+  const resolveStatus = (): FetchStatus => {
+    if (!client.isAuthenticated()) {
+      return 'unauthorized';
```
Comment (amanpdyadav) [inline src/auth/hooks.ts:155]: May be better to use enum types

File: src/auth/hooks.ts [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    [client]
+  );
+
+  useEffect(() => {
+    const autoFetch = async (): Promise<void> => {
+      if (currentStatus !== 'ready') {
+        return;
+      }
+      fetchTokens({
+        audience: String(process.env.REACT_APP_API_BACKEND_AUDIENCE),
```
Comment (amanpdyadav) [inline src/auth/hooks.ts:208]: Better to use getEnv from previous commit.

File: src/auth/index.ts [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+    if (currentUser) {
+      return Promise.resolve(currentUser);
+    }
+    if (isInitialized()) {
+      return Promise.resolve(undefined);
+    }
+    return new Promise((resolve, reject) => {
+      init()
+        .then(() => resolve(getUser()))
+        .catch(e => reject(e));
```
Comment (amanpdyadav) [inline src/auth/index.ts:391]: Can be simplified to `return init() .then(() => getUser());` from line 388 to 393

File: src/auth/oidc-react.ts [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+    const storedString = sessionStorage.getItem(userKey);
+    if (
+      !storedString ||
+      storedString.length < 2 ||
+      storedString.charAt(0) !== '{'
+    ) {
+      return undefined;
+    }
+    try {
+      return JSON.parse(storedString);
```
Comment (amanpdyadav) [inline src/auth/oidc-react.ts:117]: try {
      const parsedData = JSON.parse(storedString);
     return parsedData instance of Array ? undefined  : parsedData
} catch (e) {
      return undefined;
    }
  };

**Review by amanpdyadav — CHANGES_REQUESTED** [review]
Looks good minor changes and its good to go.

File: src/auth/oidc-react.ts (hunk 4 above)
Comment (mingfeng) [reply src/auth/oidc-react.ts:117]: Now as I look into it, I think the original solution is actually a simple and effective way to check the stored data. It's not only array that we need to rule out, but also null, boolean, string, etc. So we'll need to do something like:

```
const parsedData = JSON.parse(storedString);
if (parsedData === null || Array.isArray(parsedData) || typeof parsedData === 'boolean' || typeof parsedData === 'string'  || typeof parsedData === 'number') {
    return undefined.
}
```

**Review by amanpdyadav — APPROVED** [review]
LGTM 🚢

**Pull request merged**